test/functional/feature_discover.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test -discover command."""
+
+import socket
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+
+def is_valid_ipv4_address(address):
+    try:
+        socket.inet_aton(address)
+    except socket.error:
+        return False
+    return True
+
+
+def is_valid_ipv6_address(address):
+    try:
+        socket.inet_pton(socket.AF_INET6, address)
+    except socket.error:
+        return False
+    return True
+
+
+class DiscoverTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+
+    def validate_addresses(self, addresses_obj):
+        for address_obj in addresses_obj:
+            address = address_obj['address']
+            self.log.info(f"Validating {address}")
+            valid = (is_valid_ipv4_address(address)
+                     or is_valid_ipv6_address(address))
+            assert_equal(valid, True)
+
+    def test_local_addresses(self, test_case, should_be_empty=False):
+        self.log.info(f"Restart node with {test_case}")
+        self.restart_node(0, test_case)
+        network_info = self.nodes[0].getnetworkinfo()
+        network_enabled = [n for n in network_info['networks']
+                           if n['reachable'] and n['name'] in ['ipv4', 'ipv6']]
+        local_addrs = list(network_info["localaddresses"])
+        if should_be_empty or not network_enabled:
+            assert_equal(local_addrs, [])
+        elif len(local_addrs) > 0:
+            self.validate_addresses(local_addrs)
```

Comment:
On my system (Debian 5.16.14-1 (2022-03-15) x86_64 GNU/Linux) this is never hit and so `validate_addresses()`, `is_valid_ipv4_address()` and `is_valid_ipv6_address()` are all unused. Is it ever hit for you?

(maybe have a look at `test/functional/feature_bind_port_discover.py` and add the tests there, if helpful)

Reply:
Yes, it hits here. Example:

```
2022-03-24T17:37:48.362000Z TestFramework (INFO): Restart node with ['-listen', '-discover']
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:c84:4731:10d6:7311
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:1c3e:c153:e75f:59c3
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:3132:b83c:1915:c06e
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:4152:dcc5:5862:734
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:4d52:b4d3:d93d:8f8e
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:5594:b545:6081:a40b
2022-03-24T17:37:49.181000Z TestFramework (INFO): Validating 2804:14d:5281:8ae2:f97b:bd76:25aa:a0d3
```

I researched about `localaddresses` and noticed it's not accurate, this is a reason I don't ensure in the test that `localaddresses` MUST be filled. See: https://superuser.com/questions/1348948/cant-pipe-in-bashs-mapfile-but-why

Reply:
On which platform/OS?

Reply:
MacOS 12.0.1 Monterey

I edited the last comment because the link was wrong haha.

Reply:
Please, try again with the new changes.